datafusion/optimizer/src/type_coercion.rs

```diff
@@ -64,7 +66,15 @@ impl OptimizerRule for TypeCoercion {
             _ => DFSchemaRef::new(DFSchema::empty()),
         };
 
-        let mut expr_rewrite = TypeCoercionRewriter { schema };
+        let mut execution_props = ExecutionProps::new();
+        execution_props.query_execution_start_time =
+            optimizer_config.query_execution_start_time;
+        let const_evaluator = ConstEvaluator::try_new(&execution_props)?;
```

Comment:
We could move this initialization into a constructor to avoid doing it per plan.

Reply:
A useful suggestion, but I couldn't find a way to do this given the current trait/method signatures and making the borrow checker happy.
Given that in the expression simplifier pass one is instantiated per expression (instead of per plan), I guess this should be ok enough?

Reply:
Sure. It isn't expensive, so I am sure it is fine.

Reply:
I think the const evaluator is somewhat expensive to run (but not construct); We should keep an eye on the performance of query planning, but I think this is reasonable for now